[wasm] improve memory access and marshaling range checks

pavelsavara opened the pull request:
- capture `updateGlobalBufferAndViews`
- remove `setI64`, `getI64`
- add `setI52`, `getI52`, `getU52`, `setU52`
- add `getI64Big`, `setI64Big`
- change `bind_static_method` to use int52 with `l` signature.
- inlined JS asserts
- fixed existing binding for unsigned types
- implemented range validations for all memory access and interop

pavelsavara commented:
/azp run runtime-extra-platforms

kg commented on src/mono/wasm/runtime/exports.ts:
```diff
@@ -191,6 +193,11 @@ function initializeImportsAndExports(
     replacements.fetch = runtimeHelpers.fetch;
     replacements.readAsync = readAsync_like;
     replacements.requireOut = module.imports.require;
+    const originalUpdateGlobalBufferAndViews = replacements.updateGlobalBufferAndViews;
```
where is replacements.updateGlobalBufferAndViews initialized?

pavelsavara replied:
`src/mono/wasm/runtime/cjs/dotnet.cjs.lib.js`, that's inserted into the emscripten code

kg commented on src/mono/wasm/runtime/memory.ts:
```diff
@@ -114,3 +136,7 @@ export function getF32(offset: _MemOffset): number {
     return Module.HEAPF64[<any>offset >>> 3];
 }
+
+export function updateGlobalBufferAndViews(buffer: Buffer): void {
```
we should give this a different name from the function it's replacing, since it doesn't call the original or do what the original did

pavelsavara replied:
renamed to `afterUpdateGlobalBufferAndViews`

pavelsavara commented:
/azp run runtime-extra-platforms

kg commented on src/mono/wasm/runtime/memory.ts:
```diff
@@ -114,3 +136,7 @@ export function getF32(offset: _MemOffset): number {
 }
+
+export function afterUpdateGlobalBufferAndViews(buffer: Buffer): void {
+    HEAPI64 = new BigInt64Array(buffer);
```
we should probe whether this type is available in the browser and generate a meaningful error if it isn't

pavelsavara replied:
done

lewing commented:
/azp run runtime-wasm

pavelsavara commented on src/mono/wasm/runtime/memory.ts:
```diff
@@ -60,13 +61,23 @@ export function setI16(offset: _MemOffset, value: number): void {
+ */
 export function setI64(offset: _MemOffset, value: number): void {
-    Module.setValue(<VoidPtr><any>offset, value, "i64");
+    HEAPI64[<any>offset >>> 3] = BigInt(<number>value);
```
@kg noted that maybe bitwise math would be faster than the allocation. we should explore that.

pavelsavara replied:
done

pavelsavara committed:
* new memory accessors setI52, setI64Big, getI52, getI64Big
* removed support for long form automatic marshaler. It has impact to mono_bind_static_method

kg commented:
Big fan of the I52 / I64Big split, good idea

pavelsavara commented:
I implemented feedback from @kg about availability of `BigInt` and speed of `BigInt` allocation vs arithmetic.

I also had discussion with @lambdageek about 52 integral bits of Number vs Int64. 
There are use-cases where 52 bits are useful, for example for Date.value.

We concluded that emscripten's silent usage of only lower 32 bit is bad. 

I also realized that silent usage of lower 52bits would be similarly bad.
So I removed support for marshaling `long` signature form `bind_static_method`, which did silent 32bits till now.

This is breaking change, but failing fast should be better than silent data corruption on runtime.

pavelsavara committed:
doc

pavelsavara committed:
tests

pavelsavara commented:
We have following options: 
A) stop automatically marshaling long
B) keep it as it is - lower int32, silently ignoring overflow
C) lower int52, silently ignoring overflow
D) lower int52, throw on overflow
E) lower int32, throw on overflow
F) round to nearest floating point (convert to double)

For now I did A) but maybe for the old marshaler, B) is better choice?

For the [new marshaler](https://github.com/dotnet/runtime/pull/66304), I think we will let the user to explicitly choose between D) and BigInt. And they you always declare the parameters as double and do the conversion themselves.

pavelsavara committed:
feedback from @kg

pavelsavara committed:
fix

lewing commented:
cc @radekdoulik

kg reviewed (changes requested):
This looks great, but we should make sure the C# -> JS path handles longs correctly (I think we just need to update the unbox primitive stuff to intentionally work in terms of I52 and throw an exception when unboxing if the long is too large)

pavelsavara committed:
Merge branch 'main' into wasm_resize_and_int64

# Conflicts:
#	src/mono/wasm/runtime/memory.ts

pavelsavara committed:
- made assert silent on Release config

feedback from @kg
- test for not marshaling long from C# to JS
- negative test for marshaling NaN as long

pavelsavara commented:
I realized that all the other `setXXX` are also silently trimming bits and not throwing on overflow.
Should I introduce range check to all of them ? 
Or should I remove it from Int52 ?

pavelsavara commented:
> This looks great, but we should make sure the C# -> JS path handles longs correctly (I think we just need to update the unbox primitive stuff to intentionally work in terms of I52 and throw an exception when unboxing if the long is too large)

I think we never had `Number` -> `long` conversion on `mono_bind_static_method`. I added test for it.

pavelsavara committed:
- fixed marshaling of uint32, uint16 and byte to be unsigned
- implemented range check on all set memory operations
- implemented also uint52
- differentiated bool marshaling because it shoud have different validation and message

pavelsavara committed:
inlined asserts

pavelsavara commented:
> Should I introduce range check to all of them ?

We agreed with @kg to range-check them all. 
That forced me to fix broken marshaling of unsigned types and bool.
I also implemented uint52.
And I implemented `assert` inlining, to make sure it doesn't allocate closure.

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
fix test

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
fix message

pavelsavara commented:
/azp run runtime-wasm

pavelsavara committed:
optimize inline

kg commented on src/mono/wasm/runtime/rollup.config.js:
```diff
@@ -48,13 +48,14 @@ const format = "iife";
     // eslint-disable-next-line quotes
-    pattern: /assert\(([^,]*), *("[^"]*")\);/g,
-    replacement: "if (!($1)) throw new Error(`Assert failed: ${$2}`); // inlined assert"
+    pattern: /assert\(([^,]*), *"([^"]*)"\);/g,
```
I really think this will bite us later. Let's give it a more unique name

pavelsavara replied:
more unique name is good idea, maybe `mono_assert` ?
With that, the worst could happen is that it would not be matched and not inlined. 
We could add another dummier search for `mono_assert` as last step, if it can find some, it means the inlining didn't work for some reason and we could error on it.

kg replied:
Yeah, also if it fails to inline in a hot path that will show up in the profiler tools, so that's not an issue. Just want to avoid it accidentally mangling something else due to the word assert

pavelsavara committed:
rename mono_assert and make sure we replace them all

kg commented on src/libraries/System.Private.Runtime.InteropServices.JavaScript/src/System/Runtime/InteropServices/JavaScript/Runtime.cs:
```diff
@@ -154,12 +154,12 @@ internal static MarshalType GetMarshalTypeFromType(Type? type)
                 case TypeCode.Int32:
                     return MarshalType.INT;
+                case TypeCode.Byte:
+                case TypeCode.UInt16:
```
we should make sure that this won't cause us to read past the end of a boxed value in the heap

pavelsavara replied:
That means even more marshalers to implement, right ?

kg commented on src/mono/wasm/runtime/memory.ts:
```diff
@@ -14,6 +16,8 @@ function _ensure_allocated(): void {
     alloca_limit = <VoidPtr>(<any>alloca_base + alloca_buffer_size);
 }
 
+const is_bingint_supported = typeof BigInt !== "undefined" && typeof BigInt64Array !== "undefined";
```
```is_bigint_supported```

kg replied:
since the code works i'm fine with changing this later though

kg commented on src/mono/wasm/runtime/memory.ts:
```diff
@@ -41,33 +45,86 @@ export function _release_temp_frame(): void {
 
+function is_int_in_range(value: Number, min: Number, max: Number) {
+    mono_assert(typeof value === "number", () => `Value is not integer but ${typeof value}`);
+    mono_assert(Number.isInteger(value), "Value is not integer but float");
```
We can do ```Number.isSafeInteger(value)``` to handle the first two checks in one go, and for i52 we don't need to do additional range checks I think

pavelsavara replied:
Yes, I did it to differentiate exception message. I'm ok to make it faster.

kg commented on src/mono/wasm/runtime/rollup.config.js:
```diff
@@ -45,7 +46,19 @@ const banner_dts = banner + "//!\n//! This is generated file, see src/mono/wasm/
+        replacement: "if (!($1)) throw new Error(`Assert failed: $2`); // inlined mono_assert"
+    }, {
+        pattern: /^\s*mono_assert/gm,
+        failure: "previous regexp didn't inline all mono_assert statements"
```
good idea 👍

kg commented on src/mono/wasm/runtime/rollup.config.js:
```diff
@@ -159,4 +172,47 @@ function checkFileExists(file) {
+    };
+
+    function executeReplacement(self, code, id) {
+        // TODO use MagicString for sourcemap support
```
is this going to break all our source maps? or just for the lines/functions that got changed

pavelsavara replied:
We don't have source map anyway, because the way how emscripten concatenates multiple files.
If we want to fix it in the future, we would have pipe thru typescript->rollup->emscripten-closure compilers.

pavelsavara commented:
/azp run runtime-wasm
